vms/evm/database/blockdb/database.go

```diff
@@ -0,0 +1,528 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package blockdb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/ethdb"
+	"github.com/ava-labs/libevm/rlp"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/heightindexdb/meterdb"
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/utils/logging"
+
+	heightindexdb "github.com/ava-labs/avalanchego/x/blockdb"
+)
+
+var (
+	_ ethdb.Database = (*Database)(nil)
+	_ ethdb.Batch    = (*batch)(nil)
+
+	migratorDBPrefix = []byte("migrator")
+
+	// blockDBMinHeightKey stores the minimum block height of the
+	// height-indexed block databases.
+	// It is set at initialization and cannot be changed without
+	// recreating the databases.
+	blockDBMinHeightKey = []byte("blockdb_min_height")
+
+	errUnexpectedKey        = errors.New("unexpected database key")
+	errNotInitialized       = errors.New("database not initialized")
+	errAlreadyInitialized   = errors.New("database already initialized")
+	errInvalidEncodedLength = errors.New("invalid encoded length")
+)
+
+// Key prefixes for block data in [ethdb.Database].
+// This is copied from libevm because they are not exported.
+// Since the prefixes should never be changed, we can avoid libevm changes by
+// duplicating them here.
+var (
+	evmHeaderPrefix    = []byte("h")
+	evmBlockBodyPrefix = []byte("b")
+	evmReceiptsPrefix  = []byte("r")
+)
+
+const (
+	hashDataElements = 2
+	blockNumberSize  = 8
+	blockHashSize    = 32
```

Comment:
(fyi) To better explain my rationale for removing the other constants, I think these are useful because they couple multiple usage sites that all MUST match.